connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java

```diff
@@ -144,15 +144,25 @@ public boolean awaitStop(long timeoutMs) {
         }
     }
 
+    /**
+     * Remove all metrics published by this task.
+     */
+    public void removeMetrics() {
+        try {
+            removeAdditionalMetrics();
+        } finally {
+            taskMetricsGroup.close();
+        }
+    }
+
     protected abstract void execute();
 
     protected abstract void close();
 
     /**
-     * Method called when this worker task has been completely closed, and when the subclass should clean up
-     * all resources.
+     * Remove any metrics specific to the subclass. Invoked as part of {@link #removeMetrics}.
      */
-    protected abstract void releaseResources();
+    protected abstract void removeAdditionalMetrics();
```

Comment:
I am not sure I understand the need for `WorkerTask` and `WorkerSource/SinkTask` to implement different methods. Usually, this would be done by the subclasses overriding the parent class implementation and making sure to call the parent implementation somewhere with `super.xxx()`. Am I missing some requirement here that necessitates the separate methods?

Reply:
Yeah, I was thinking about this. Not super attached to either approach, just went with the one that reduced code duplication and took shared logic (try to release source/sink-specific metrics, then release generic task metrics even if that failed) and put it in one place, the abstract class.

I think an alternative might look like this in `WorkerTask`:
```java
public void removeMetrics() {
    taskMetricsGroup.close();
}
taskMetricsGroup.close();
```

and this in `WorkerSinkTask` (with something similar in `WorkerSourceTask`):
```java
@Override
public void removeMetrics() {
    try {
        sinkTaskMetricsGroup.close();
    } finally {
        super.removeMetrics();
    }
}
```

If this looks better, or you were thinking of something completely different, LMK.

Reply:
Yeah, this was exactly what I had in mind. Conceptually, both child and parent classes need to do some stuff for cleanup. Child class does its cleanup and then just delegates to parent implementation to do parent cleanup. Took me a while to realize that the separate methods on the `WorkerTask` and `WorkerSink/SourceTask` were doing exactly this. I think following the common idiom would be a bit more readable.

Reply:
Ack, will update :)